### PR TITLE
Remove CWalletTx merging logic from AddToWallet

### PR DESCRIPTION
**Pull request by ryanofsky**
This is a pure refactoring, no behavior is changing.

Instead of AddToWallet taking a temporary CWalletTx object and then potentially merging it with a pre-existing CWalletTx, have it take a callback so callers can update the pre-existing CWalletTx directly.

This makes AddToWallet simpler because now it is only has to be concerned with saving CWalletTx objects and not merging them.

This makes AddToWallet calls clearer because they can now make direct updates to CWalletTx entries without having to make temporary objects and then worry about how they will be merged.

Motivation for this change came from the bumpfee PR #8456 where we wanted to be able to call AddToWallet to make a simple update to an existing transaction, but were reluctant to, because the existing CWalletTx merging logic did not apply and seemed dangerous try to update as part of that PR. After this refactoring, the bumpfee PR could call AddToWallet safely instead of implementing a duplicate AddToWallet function.

This also allows getting rid of the CWalletTx copy constructor to prevent unintentional copying.

### CODE REVIEW

**Comment by ryanofsky** [comment]
Rebased ed457a39ebaa672cce21622d3af422fe50730580 -> b1ac3cff00a51c0b8965b7a85fa489c98409ddff to resolve merge conflicts with bumpfee (#8456).

**Comment by ryanofsky** [comment]
I'm thinking of splitting this up into two commits to make it easier to review. First commit would change CreateTransaction to return a CTransactionRef instead of CWalletTx. Second commit would change AddToWallet to accept a CTransactionRef instead of a CWalletTx.

If any reviewers would prefer this you can let me know.

**Comment by TheBlueMatt** [comment]
Splitting commits into logical breaks is always appreciated.

**Comment by TheBlueMatt** [comment]
Hum, reading this (finally) now, I'm not sure if I'm a big fan of this approach. The call-function-which-callsback-to-let-you-fill-in-things-that-really-should-have-been-arguments pattern really sucks. On the flip side, passing in CWalletTxes, and copying those into mapWallet can also lead to issues if you try to do anything to the object you just passed in thinking it is what got stored in the wallet (a mistake I made recently).

Maybe we should adapt some of these functions (AddToWallet/CommitTransaction/whatever) to just always take an rvalue to a CWalletTx. This leaves only AddToWalletIfInvolvingMe (I think) which has to update-or-add, and that can just use an internal version of AddToWallet which takes the CWalletDB as an argument instead of creating its own.

**Comment by TheBlueMatt** [comment]
I'm just generally not a fan of callbacks making ownership models inconsistent. eg this kind of thing is where people always fuck up lockordering (though admittedly less so in this particular case, more the general case of callbacks going in both directions between modules).

So I do like this version much better, but still not sure the call back is required to be publicly exposed...Can we just leave MarkReplaced and add a similar function for importpruned funds to use (I dont believe the double-disk-sync in importprunedfunds will make for inconsistent/bad on-disk state?). Then we can make AddToWallet(..., callback) private and use AddToWalletIfInvolvingMe (which should probably also take a CTransactionRef, not a CTransaction, though maybe thats for a separate PR) publicly.

**Comment by ryanofsky** [comment]
> Can we just leave MarkReplaced and add a similar function for importpruned funds to use

Will try that. In the meantime I moved the non-AddToWallet cleanup to #9680 so it can be considered separately. By the way #9369 is another PR which significantly simplifies AddToWallet.

**Comment by ryanofsky** [comment]
> Can we just leave MarkReplaced and add a similar function for importpruned funds to use

Implemented this suggestion. (New function is `CWallet::AddTransaction`). I also went ahead and disabled the CWalletTx copy constructor. Interestingly this uncovered a bunch of cases where we were unwittingly copying wallet transactions in loops due to writing:
```c++
for (const std::pair<uint256, CWalletTx>& pairWtx : pwallet->mapWallet)
```

instead of:
```c++
for (const std::pair<const uint256, CWalletTx>& pairWtx : pwallet->mapWallet)
```

**Comment by MarcoFalke** [comment]
Needs rebase to fix travis (sorry)

**Comment by ryanofsky** [comment]
> Needs rebase to fix travis (sorry)

Rebased 71bce6b0f4aee487026afd31ada41a4e6028abd3 -> 31297f007b2424ab5036258ed782aacc70237a7f ([pr/atw-nomerge.33](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.33) -> [pr/atw-nomerge.34](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.34))

**Review by kallewoof — COMMENTED** [review]
Light utACK 59ba4deec713669504d284b415bd31f049866f4c

**Review by ryanofsky — COMMENTED** [review]
@kallewoof, this is based on #9680, so perhaps you could add your ACK there, since it looks like you reviewed commits from that PR.

---

Added 1 commits 31297f007b2424ab5036258ed782aacc70237a7f -> 64f182fa587f35081ed8026b44e2a36010364153 ([pr/atw-nomerge.34](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.34) -> [pr/atw-nomerge.35](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.35), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/atw-nomerge.34...pr/atw-nomerge.35))
Squashed 64f182fa587f35081ed8026b44e2a36010364153 -> 01ab386a6957de4b97ed52f2b1b2d45aa7ba4e0f ([pr/atw-nomerge.35](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.35) -> [pr/atw-nomerge.36](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.36))
Rebased 01ab386a6957de4b97ed52f2b1b2d45aa7ba4e0f -> e15246fc310599770332021d14b96ce3f213a82b ([pr/atw-nomerge.36](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.36) -> [pr/atw-nomerge.37](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.37)) due to conflict with #12658

**Comment by kallewoof** [comment]
Re-slight-utACK 8a6ec70

**Comment by ryanofsky** [comment]
Rebased e15246fc310599770332021d14b96ce3f213a82b -> d62357a7c7c8c34a55580a31ddbfd04a9fa72591 ([pr/atw-nomerge.37](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.37) -> [pr/atw-nomerge.38](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.38)) due to conflict with #12681 and after merge of base PR #9680.

File: src/wallet/wallet.h
Comment (promag) [inline src/wallet/wallet.h]: In commit "Avoid copying CWalletTx in LoadToWallet",

This and the `Init()` change could be in a different (trivial) commit.

File: src/wallet/test/wallet_tests.cpp
Comment (promag) [inline src/wallet/test/wallet_tests.cpp]: In commit "Remove CWalletTx merging logic from AddToWallet",

Should stay? Related to #12681.

Looks like [this code](https://github.com/bitcoin/bitcoin/pull/12681/files#diff-b2bb174788c7409b671c46ccc86034bdR3816) is not hit, otherwise the lock held assertion in `LookupBlockIndex` would cause travis failure here?

**Review by promag — COMMENTED** [review]
> I'm just generally not a fan of callbacks

+1

The merging code could still be removed but instead of callbacks, the caller would first lookup the transaction and if it exists, it would update accordingly, otherwise add it.

Anyway, this sounds better than the current implementation.

> Interestingly this uncovered a bunch of cases where we were unwittingly copying wallet transactions in loops due to writing:

Use `auto` there?

**Review by ryanofsky — COMMENTED** [review]
Split up commits d62357a7c7c8c34a55580a31ddbfd04a9fa72591 -> 65a15b7752e0ca0af3c381a26b07833a81523d30 ([pr/atw-nomerge.38](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.38) -> [pr/atw-nomerge.39](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.39)) as suggested
https://github.com/bitcoin/bitcoin/pull/9381#discussion_r174958125, no other code changes.
Rebased 65a15b7752e0ca0af3c381a26b07833a81523d30 -> aa967a383a491ac2f90becaf998e91ac684b2d1f ([pr/atw-nomerge.39](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.39) -> [pr/atw-nomerge.40](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.40)) due to conflict with #12891
Rebased aa967a383a491ac2f90becaf998e91ac684b2d1f -> 1d6c0ba1159d67c7b6dec8891e41fc20ec66370e ([pr/atw-nomerge.40](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.40) -> [pr/atw-nomerge.41](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.41)) due to conflict with #11851
Rebased 1d6c0ba1159d67c7b6dec8891e41fc20ec66370e -> cad8b14857d8e310ebad8a850c9e6450e13360fc ([pr/atw-nomerge.41](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.41) -> [pr/atw-nomerge.42](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.42)) due to conflict with #13081

---

>> I'm just generally not a fan of callbacks
>
> +1
>
> The merging code could still be removed but instead of callbacks, the caller would first lookup the transaction and if it exists, it would update accordingly, otherwise add it.

Callbacks can be used badly, and callbacks can also be used well. I think it's better to look at advantages and disadvantages of using them in specific cases than to talk about callbacks in general. The reason I think callbacks make sense here is that `AddToWallet` gives callers a way to add and update transactions to the database, and takes care of the locking, serialization, i/o, and notification on callers behalf. The callback gives `AddToWallet` just enough flexibility to allow its callers to be able to mutate the `CWalletTx` object in the ways they need, without having to handle the other low level details.

It seems to me like the suggestion above essentially amounts to inlining `AddToWallet` in the various places it is called, which would be more verbose and complicated for callers. But if there are advantages to doing this, I'd be interested to know about them.

---

>> Interestingly this uncovered a bunch of cases where we were unwittingly copying wallet transactions in loops due to writing:

> Use `auto` there?

Using `auto` would avoid the copying in these particular cases, but disabling the copy constructor avoids it in all cases, and prevents new unintended copies in the future.

**Comment by ryanofsky** [comment]
> Needs rebase 

Rebased cad8b14857d8e310ebad8a850c9e6450e13360fc -> a494d2a4472ea63620b1f6cce20cb726590eb75b ([pr/atw-nomerge.42](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.42) -> [pr/atw-nomerge.43](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.43)) due to minor conflict with #13437.

File: src/wallet/wallet.cpp
Comment (practicalswift) [inline src/wallet/wallet.cpp]: Typo found by `codespell`: tranaction

File: src/wallet/wallet.cpp
Comment (practicalswift) [inline src/wallet/wallet.cpp]:
```
2018-09-20 06:04:29 clang-tidy(pr=9381): src/wallet/wallet.cpp:843:64: warning: the parameter 'update_wtx' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param]
2018-09-20 06:04:29 clang-tidy(pr=9381): src/wallet/wallet.cpp:903:66: warning: the parameter 'update_wtx' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param]
```

File: src/wallet/test/wallet_tests.cpp
Comment (practicalswift) [inline src/wallet/test/wallet_tests.cpp]: Seems like `SetMerkleBranch` can be removed.

```
2018-09-20 06:29:02 cppcheck(pr=9381): [src/wallet/wallet.cpp:4237]: (style) The function 'SetMerkleBranch' is never used.
```

File: src/wallet/walletdb.cpp
Comment (practicalswift) [inline src/wallet/walletdb.cpp]: `new_tx` not used?

File: src/wallet/wallet.cpp
Comment (practicalswift) [inline src/wallet/wallet.cpp]: `new_tx` not used?

File: src/wallet/wallet.cpp
Comment (practicalswift) [inline src/wallet/wallet.cpp]: `new_tx` not used?

**Review by meshcollider — COMMENTED** [review]
Light utACK https://github.com/bitcoin/bitcoin/pull/9381/commits/8cbc44f92f0932c87abadbd7c85e246d5b37b707

Just a couple of small questions inline

**Review by Sjors — APPROVED** [review]
@achow101 how does this jive with wallet boxes #16341?

Code review light-ACK b8f3870c6a46052a80c694c0744a3eb56a9116ac (some things go over my head, but do look cleaner). Tested sending coins between wallets, RBF and then abandoning the old one and waiting for confirmation.

**Review by ryanofsky — COMMENTED** [review]
Thanks for the review!

Updated b8f3870c6a46052a80c694c0744a3eb56a9116ac -> 825a514481dd69dc7038c7f4c974a76cb2becd87 ([pr/atw-nomerge.61](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.61) -> [pr/atw-nomerge.62](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.62), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/atw-nomerge.61..pr/atw-nomerge.62)) adding a update_wtx comment.

> @achow101 how does this jive with wallet boxes #16341?

Not achow, but I've been slogging through #16341, and I think there is not too much overlap since this PR is concerned with transactions, while the other is concerned with scripts and keys. I definitely think this PR has much less priority than #16341 and shouldn't interfere with it. I've been rebasing this PR a while anyway and it's been easy to maintain.

**Comment by achow101** [comment]
> @achow101 how does this jive with wallet boxes #16341?

AFAICT They don't interact at all. The only part of #16341 that touches transactions is a small section in `AddToWalletIfInvolvingMe` which this PR does not touch.

**Comment by Sjors** [comment]
re-ACK 825a514481dd69dc7038c7f4c974a76cb2becd87, but Travis is in a state of despair, perhaps due to #16582 (@MarcoFalke?)

File: src/wallet/wallet.cpp
Comment (ariard) [inline src/wallet/wallet.cpp]: Hmmm new execution flow doesn't make it easier to understand. Maybe mapWallet should be access directly in `ReadKeyValue`, when as you said in another comment wallet components are broken up in smaller classes. I think, but need to test, that `MarkConflicted` and its loop is useless here.

**Review by ariard — COMMENTED** [review]
Concept ACK, did only a light code review, I think changes are going in the right way to let caller manage the scope of transactions changes, at same time it points way in which CWallet could be refactored in better way.

It's overlap with #9381 but rebased in any order shouldn't be ugly

**Review by ryanofsky — COMMENTED** [review]
Rebased cb4fce1f5c4b83d08704a0423f88add021405aae -> 47d95638c682d8efd1e46dfcb300b75aa970f7a0 ([pr/atw-nomerge.63](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.63) -> [pr/atw-nomerge.64](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.64)) due to conflict with #16624

**Comment by meshcollider** [comment]
Sorry about all the rebases, Concept ACK

**Comment by laanwj** [comment]
Concept ACK

**Comment by ryanofsky** [comment]
Rebased dff3c12515773bef07756bb2420da5a1ac8e1438 -> 93510bcfe80385203880d1211cb94f621411839e ([`pr/atw-nomerge.74`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.74) -> [`pr/atw-nomerge.75`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.75), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/atw-nomerge.74-rebase..pr/atw-nomerge.75)) due to minor conflict with #13339

File: src/wallet/walletdb.cpp
Comment (Sjors) [inline src/wallet/walletdb.cpp]: In 754eb2112aac4729805dcc547bef1de956c3c14d `Avoid copying CWalletTx in LoadToWallet`

Maybe add a comment here that `LoadToWallet` creates a fresh `CWalletTx` and that we fill it with `ssValue` (hence my rename to `fill_wtx` suggestion above).

Since a wallet doesn't (shouldn't? otherwise add `assert(new_tx)`) have duplicate transactions `new_tx` should be `true`, but if some reason a wallet entry does exists it will be overridden.

File: src/wallet/wallet.cpp
Comment (Sjors) [inline src/wallet/wallet.cpp]: Maybe rename `update_wtx` to `fill_wtx `.

File: src/wallet/walletdb.cpp
Comment (Sjors) [reply src/wallet/walletdb.cpp]: Maybe briefly mention this in the commit message, as the change from `std::vector` to `std::list` looks a bit out of the blue (until you try compiling without).

**Review by Sjors — APPROVED** [review]
ACK 93510bcfe80385203880d1211cb94f621411839e

**Review by ryanofsky — COMMENTED** [review]
Thanks for reviewing!

Updated 93510bcfe80385203880d1211cb94f621411839e -> fc4df9dc2910c4a57d5e81573faf7952e1c3ee42 ([`pr/atw-nomerge.75`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.75) -> [`pr/atw-nomerge.76`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.76), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/atw-nomerge.75..pr/atw-nomerge.76)) with suggestions and a few tweaks to make diffs a little smaller

**Comment by ryanofsky** [comment]
Rebased fc4df9dc2910c4a57d5e81573faf7952e1c3ee42 -> f1fc7892608fb258da86017c79ffb4b1a477014a ([`pr/atw-nomerge.76`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.76) -> [`pr/atw-nomerge.77`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.77), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/atw-nomerge.76-rebase..pr/atw-nomerge.77)) due to conflict with #18278

**Comment by Sjors** [comment]
utACK f1fc7892608fb258da86017c79ffb4b1a477014a

**Comment by ryanofsky** [comment]
Rebased f1fc7892608fb258da86017c79ffb4b1a477014a -> ece5d1cf46cf8c335dee4bb6affe514d7b33f306 ([`pr/atw-nomerge.77`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.77) -> [`pr/atw-nomerge.78`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.78), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/atw-nomerge.77-rebase..pr/atw-nomerge.78)) due to conflict with #17954

**Comment by Sjors** [comment]
re-utACK ece5d1c (rebased)

**Comment by ryanofsky** [comment]
Rebased ece5d1cf46cf8c335dee4bb6affe514d7b33f306 -> f2892d8eb5404a34f3b09d3b9ac65d58c965e362 ([`pr/atw-nomerge.78`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.78) -> [`pr/atw-nomerge.79`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.79), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/atw-nomerge.78-rebase..pr/atw-nomerge.79)) due to conflict with #16426

**Comment by Sjors** [comment]
Light re-utACK f2892d8eb5404a34f3b09d3b9ac65d58c965e362, looks like a reasonably straight-forward rebase.

**Review by meshcollider — COMMENTED** [review]
utACK f2892d8eb5404a34f3b09d3b9ac65d58c965e362

File: src/wallet/wallet.cpp
Comment (MarcoFalke) [inline src/wallet/wallet.cpp:3010]: Does the `std::move` still work after going through a lambda capture by reference?

File: src/wallet/wallet.cpp
Comment (MarcoFalke) [inline src/wallet/wallet.cpp:3018]: Doesn't this crash line crash the node when the wallet can not write the tx to disk?

**Review by maflcko — APPROVED** [review]
ACK f2892d8

Just some questions

File: src/wallet/wallet.cpp
Comment (sipa) [reply src/wallet/wallet.cpp:3010]: Yes. mapValue is a reference to the same thing it was outside of the lambda. std::move casts it to an rvalue reference.

File: src/wallet/wallet.cpp
Comment (ryanofsky) [reply src/wallet/wallet.cpp]: re: https://github.com/bitcoin/bitcoin/pull/9381#discussion_r420187836

> Doesn't this line crash the node when the wallet can not write the tx to disk?

Wow, good catch! I'm not sure current behavior is ideal either but it wasn't my intention to change it. Updated PR

File: src/wallet/wallet.cpp
Comment (MarcoFalke) [reply src/wallet/wallet.cpp:3010]: Oh, this has nothing to do with lambdas. I didn't know that references (as long as they are not const) can be moved.

Comment (ryanofsky) [reply src/wallet/wallet.cpp:3010]: re: https://github.com/bitcoin/bitcoin/pull/9381#discussion_r420187394

> Oh, this has nothing to do with lambdas. I didn't know that references (as long as they are not const) can be moved.

Yep, `std::move` isn't affected by outside things like this. `std::move(mapValue)` is only a type cast from `mapValue_t&` to `mapValue_t&&` that makes the compiler favor the `operator=(mapValue_t&&)` assignment overload instead of the `operator=(const mapValue_t&)` one for `wtx.mapValue =`

Comment (sipa) [reply src/wallet/wallet.cpp:3010]: Even const ones, actually - though unless they have mutable fields, there is no effect.

**Review by ryanofsky — COMMENTED** [review]
Updated f2892d8eb5404a34f3b09d3b9ac65d58c965e362 -> 28b112e9bd3fd1181c0720306051ba7efca8b43 ([`pr/atw-nomerge.79`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.79) -> [`pr/atw-nomerge.80`](https://github.com/ryanofsky/bitcoin/commits/pr/atw-nomerge.80), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/atw-nomerge.79..pr/atw-nomerge.80)) reverting some CommitTransaction changes to avoid changing behavior and crashing on failure

**Comment by MarcoFalke** [comment]
Ah sorry, no need to revert your changes. If the wallet can't write to disk, a simple `CHECK_NONFATAL(wtx)` should be sufficient

File: src/wallet/wallet.cpp
Comment (MarcoFalke) [reply src/wallet/wallet.cpp:3010]: With "no effect" you mean `const&` will stay `const&`, right? See also the example, which does not compile, because the `const&` copy constructor is deleted:

```cpp
#include <memory>

struct Test{
    const std::unique_ptr<int>b;
    Test(const std::unique_ptr<int>& a):b{std::move(a)}{}
};

Comment (sipa) [reply src/wallet/wallet.cpp:3010]: `std::move(a)` where a is `const T&`, returns something of type `const T&&`. If T has for example a `operator=(const T&&)`, then that one would be selected in assignment. If there is no such operator, `operator=(const T&` will be selected instead. In general a `const T&&` assignment/constructor operator only makes sense if a class has mutable fields.

Comment (MarcoFalke) [reply src/wallet/wallet.cpp:3010]: TIL that `const T&&` exists.

**Comment by ryanofsky** [comment]
> Ah sorry, no need to revert your changes. If the wallet can't write to disk, a simple `CHECK_NONFATAL(wtx)` should be sufficient

I don't think CHECK_NONFATAL would be right, because this is checking a runtime error, not an internal assertion about how code works. I can go back to the earlier version and add this if that's your preference. But actually I like the update better because it makes the overall diff smaller and avoids changing behavior in a refactoring PR. I think ideally if different error handling is needed here, it would be handled in a standalone PR with a unit test.

**Comment by MarcoFalke** [comment]
Well, our assumption in the code is that the wallet can write to disk. If that assumption is violated it seems fine to assert or do that `CHECK` thingy. But I also see your point to make it a separate runtime error.

Anyway, re-ACK 28b112e9bd3fd1181c0720306051ba7efca8b436

**Comment by meshcollider** [comment]
re-utACK 28b112e9bd3fd1181c0720306051ba7efca8b436

**Pull request merged**
